GenderPayGap.WebUI/Controllers/Account/ChangeEmailController.cs

```diff
@@ -0,0 +1,202 @@
+            }
+
+            User user = dataRepository.Get<User>(changeEmailToken.UserId);
+
+            var viewModel = new VerifyEmailChangeViewModel
+            {
+                User = user,
+                Code = code,
+                NewEmailAddress = changeEmailToken.NewEmailAddress
+            };
+            return View("VerifyEmail", viewModel);
+        }
+
+        // The user has typed in their password
+        // We validate the email verify token, and check that their password is correct
+        // We then go ahead and change their email address
+        [AllowAnonymous]
+        [HttpPost("verify-email")]
+        public IActionResult VerifyEmailPost(VerifyEmailChangeViewModel viewModel)
+        {
+            ChangeEmailVerificationToken changeEmailToken = Encryption.DecryptModel<ChangeEmailVerificationToken>(viewModel.Code);
+
+            if (TokenHasExpired(changeEmailToken))
+            {
+                string error = "Your email verification link has expired. Please go to My Account and start the email change process again.";
+                return View("VerifyEmailError", error);
+            }
+
+            User user = dataRepository.Get<User>(changeEmailToken.UserId);
+            viewModel.User = user;
+            viewModel.NewEmailAddress = changeEmailToken.NewEmailAddress;
+
+            // Check if the user has entered a password (they might have left this field blank)
+            viewModel.ParseAndValidateParameters(Request, m => m.Password);
+            if (viewModel.HasAnyErrors())
+            {
+                return View("VerifyEmail", viewModel);
+            }
+
+            if (!userRepository.CheckPasswordAsync(user, viewModel.Password).Result)
+            {
+                viewModel.AddErrorFor(m => m.Password, "Incorrect password");
+                return View("VerifyEmail", viewModel);
+            }
+
+            if (OtherUserWithThisEmailAddressAlreadyExists(viewModel.NewEmailAddress))
+            {
+                string error = "This email address is already taken by another account.";
+                return View("VerifyEmailError", error);
+            }
+
+            string oldEmailAddress = user.EmailAddress;
+
+            userRepository.UpdateEmail(user, changeEmailToken.NewEmailAddress);
+
+            NotifyBothOldAndNewEmailAddressesThatEmailAddressHasBeenChanged(oldEmailAddress, changeEmailToken.NewEmailAddress);
+
+            return View("ChangeEmailComplete", changeEmailToken.NewEmailAddress);
+        }
+
+        private static bool TokenHasExpired(ChangeEmailVerificationToken changeEmailToken)
+        {
+            DateTime verifyExpiryDate = changeEmailToken.TokenTimestamp.AddDays(Global.EmailVerificationExpiryDays);
+            return verifyExpiryDate < VirtualDateTime.Now;
+        }
+
+        private void NotifyBothOldAndNewEmailAddressesThatEmailAddressHasBeenChanged(string newOldAddress, string newEmailAddress)

```

Comment:
could the first argument be `oldAddress` instead?

Reply:
Oops. Yes, I meant for this to be `oldEmailAddress` and `newEmailAddress`